### ISSUE TITLE
Specifying `resources` kwarg in client.submit() prevents the job from completing

### ISSUE THREAD

**Issue by rweisman-efp** [dask/dask#9434]
**What happened**:
- The future is put into the "pending" state and remains that way indefinitely.
- Consequently, the call to `gather()` never returns.
- No mentions of the job's existence appear in the web UI.
- This behavior seems consistent regardless of how the "MEMORY" resource is configured on client workers (>20, <20, or not at all)

**What you expected to happen**:
If "MEMORY" is a configured resource >=20 on at least one worker:
- The job is assigned to that worker(s)
- The future is put into a completed state shortly thereafter
- The future's result is 42

Otherwise:
- The future is cancelled as the current state of the cluster will never be able to process the task given the resources available

**Minimal Complete Verifiable Example**:

```python
from dask.distributed import Client

def some_task():
    return 42

dask_client = Client()
job = dask_client.submit(some_task, resources={'MEMORY': 20})
dask_client.gather(job)
print(job.result())
```

**Anything else we need to know?**:
- Tried as well using dask annotations with same result
- The issue persists even when requesting `resources={'MEMORY': 0}`
- If you remove the "resources" kwarg entirely, this works as expected

**Attempts**

Tried running against the local client as shown above, with no "MEMORY" resource configured:
- The most recent log was:
> distributed.scheduler - INFO - Receive client connection: Client-320f56ab-24b9-11ed-aaed-0233a8571446
  - The cluster dump reports receipt of the task:
```
<TaskState 'some_task-b97c61448795b10cddb590d67d5cf67b' no-worker>
```

Tried running against a cluster, whose workers were launched with the environment variable `DASK_DISTRIBUTED__WORKER__RESOURCES__MEMORY=32` per [the documentation](https://distributed.dask.org/en/stable/resources.html):
- Identical symptoms

**Environment**:

- First try on Dask version: 2022.6.0, Python version: 3.9.7
- Second try on Dask version: 2022.7.0, Python version: 3.10.4 (default install to new conda env on this distro as of writing)
- Operating System: Ubuntu 20.04.4 LTS on AWS (GNU/Linux 5.13.0-1021-aws x86_64)
- Install method (conda, pip, source): conda

<details>
<summary>Cluster Dump State:</summary>

</details>

**Comment by pavithraes** [comment]
@rweisman-efp Welcome!

> Tried running against a cluster, whose workers were launched with the environment variable DASK_DISTRIBUTED__WORKER__RESOURCES__MEMORY=32

You're right in doing this -- because you would first need to set this "MEMORY" resource to the workers before using it in `client.submit()`.

Just to confirm, did you connect your workers+scheduler to the client?

Note that `dask_client = Client()` will create a _new cluster_ and connect a client to it. 

Could you please try:
- [Launching your scheduler and worker from the command line](https://docs.dask.org/en/stable/deploying-cli.html#command-line). And while creating the worker, you can specify the resource using environment variables
- In your code snippet, use `dask_client = Client(address=<scheduler-address>)` where "scheduler address" is the address displayed while launching your scheduler.

And let us know if it still doesn't work?

**Comment by rweisman-efp** [comment]
Hello and thanks for the prompt reply!

Originally tried against both a new cluster (`dask_client = Client()`) as well as an extant, remote cluster with that environment variable set (i.e. `dask_client = Client(address='remote.host.with.scheduler:8786')`).

Given your reply, it seems most likely the latter was misconfigured.  I'll try to launch the scheduler locally as you suggest and report back.

**Comment by rweisman-efp** [comment]
Unfortunately, the behavior is the same.  Perhaps there's a typo in how I'm specifying the environment variable?

Scheduler output:
```
$dask-scheduler
2022-08-31 16:58:37,533 - distributed.scheduler - INFO - -----------------------------------------------
2022-08-31 16:58:38,133 - distributed.scheduler - INFO - State start
2022-08-31 16:58:38,135 - distributed.scheduler - INFO - -----------------------------------------------
2022-08-31 16:58:38,135 - distributed.scheduler - INFO - Clear task state
2022-08-31 16:58:38,136 - distributed.scheduler - INFO -   Scheduler at: tcp://172.31.175.176:8786
2022-08-31 16:58:38,136 - distributed.scheduler - INFO -   dashboard at:                     :8787
2022-08-31 16:58:41,242 - distributed.scheduler - INFO - Register worker <WorkerState 'tcp://127.0.0.1:33659', status: init, memory: 0, processing: 0>
2022-08-31 16:58:41,251 - distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:33659
2022-08-31 16:58:41,251 - distributed.core - INFO - Starting established connection
2022-08-31 16:58:56,233 - distributed.scheduler - INFO - Receive client connection: Client-333c596a-294e-11ed-bec0-0233a8571446
2022-08-31 16:58:56,234 - distributed.core - INFO - Starting established connection
2022-08-31 16:59:15,963 - distributed.scheduler - INFO - Receive client connection: Client-3f0166d6-294e-11ed-bec0-0233a8571446
2022-08-31 16:59:15,964 - distributed.core - INFO - Starting established connection
```

Worker output:
```
$DASK_DISTRIBUTED__WORKER__RESOURCES__MEMORY=32 dask-worker tcp://127.0.0.1:8786
2022-08-31 16:58:40,539 - distributed.nanny - INFO -         Start Nanny at: 'tcp://127.0.0.1:38765'
2022-08-31 16:58:41,232 - distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:33659
2022-08-31 16:58:41,232 - distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:33659
2022-08-31 16:58:41,232 - distributed.worker - INFO -          dashboard at:            127.0.0.1:42749
2022-08-31 16:58:41,232 - distributed.worker - INFO - Waiting to connect to:       tcp://127.0.0.1:8786
2022-08-31 16:58:41,232 - distributed.worker - INFO - -------------------------------------------------
2022-08-31 16:58:41,232 - distributed.worker - INFO -               Threads:                         16
2022-08-31 16:58:41,232 - distributed.worker - INFO -                Memory:                 501.73 GiB
2022-08-31 16:58:41,232 - distributed.worker - INFO -       Local Directory: ~/dask-worker-space/worker-nym66clc
2022-08-31 16:58:41,232 - distributed.worker - INFO - -------------------------------------------------
2022-08-31 16:58:41,251 - distributed.worker - INFO -         Registered to:       tcp://127.0.0.1:8786
2022-08-31 16:58:41,251 - distributed.worker - INFO - -------------------------------------------------
2022-08-31 16:58:41,252 - distributed.core - INFO - Starting established connection
```

Result (tried with both MEMORY set to 0 and to 5):
![image](https://user-images.githubusercontent.com/103955597/187737726-d50e6389-f1bb-48da-ad4d-ebc483d97c27.png)

**Comment by ncclementi** [comment]
@rweisman-efp I'm looking at the docs on how to specify worker resources [here](https://distributed.dask.org/en/stable/resources.html) and I see that the scheduler has the resources set too. see that in the example in the docs there is this step:

```
dask-worker scheduler:8786 --resources "GPU=2"
dask-worker scheduler:8786 --resources "GPU=2"
dask-worker scheduler:8786 --resources "MEMORY=100e9" #notice the resource set here.
```

before 

```
from distributed import Client
client = Client('scheduler:8786') #using the scheduler with resources set above

data = [client.submit(load, fn) for fn in filenames]
processed = [client.submit(process, d, resources={'GPU': 1}) for d in data]
final = client.submit(aggregate, processed, resources={'MEMORY': 70e9})

```

Comparing that with your example, what's missing is specifying the resources when creating the workers with the LocalCluster. Modifying your script as this runs:

```python
from dask.distributed import Client

def some_task():
    return 42

dask_client = Client(resources={"MEMORY": 20}) #notice this here
job = dask_client.submit(some_task, resources={'MEMORY': 20})
dask_client.gather(job)
print(job.result())
```

But I can see why this might be confusing from the docs. Do you think that adding an example like yours to the docs  would be more useful, if so would you be willing to open a PR?

**Comment by jrbourbeau** [comment]
Hrm, while I agree the examples @ncclementi are totally correct. One thing I noticed when playing around with the environment variable approach is that `DASK_DISTRIBUTED__WORKER__RESOURCES__MEMORY=32` sets a resource of `memory=32` (note `memory` is lower case). If I use this env var, and update @rweisman-efp's original example to use lower case `memory` in the `client.submit(..., )` call, then things work as expected

**Comment by ncclementi** [comment]
@jrbourbeau Do you mean `job = dask_client.submit(some_task, resources={'memory': 20})`. Is that setting the limit on all the workers, or is it restricting the just `some_task` . I guess what's the difference between lower and uppercase? 

This seems a bit confusing for users. Is there a way we can make the docs better

**Comment by jrbourbeau** [comment]
> Do you mean `job = dask_client.submit(some_task, resources={'memory': 20})`

Yeah, exactly. Resources are set when we create the worker with this environment variable

```
DASK_DISTRIBUTED__WORKER__RESOURCES__MEMORY=32 dask-worker
```

and then with 

```python
job = dask_client.submit(some_task, resources={'memory': 20})
```

we're saying that whenever we run the `some_task` function on the cluster, it will require 20 `memory` resources. 

> This seems a bit confusing for users. Is there a way we can make the docs better?

Agreed. Perhaps add a note to the "From environment variables" section in https://distributed.dask.org/en/stable/resources.html? 

Alternatively, we could make resources case-insensitive. Would want to think more about whether or not that's actually a good idea though...
